### PR TITLE
feat: Make UI route segments guardable through dedicated guard functions

### PR DESCRIPTION
**Pull request by glenntws**
**This PR is in proposal state and does not yet implement all necessary changes (i.e. tests) !**

Implemented the feature which I suggested as an idea yesterday (https://github.com/vercel/next.js/discussions/60560).

## What?

- This PR adds a new file type called `guard.js`.
- `guard.js` files `export default` a guard function which is executed before rendering a page.
- guard functions are nested, but inverse in their execution order compared to layouts/pages/templates (parent route segment guard executed first).
- guard functions are called in a new internal react component and thus executed in the react rendering environment (having all functionality of RSC's, but don't return anything).
- guard functions have access to route params.
- guard functions are only executed in valid routes. in other words: they are immediately executed before the rendering of a page component happens.

next-page-guard-proposal

## Why?

- In my own project, I need to be able to do sophisticated route guarding that also depends on server-side-fetched info relying on route params.
- This fetching is tedious when doing it in every page of a segment, leading to huge amount of redundancy when the segment page tree get's a little bit bigger.
- Layout files can't and logically speaking shouldn't be used for such purposes, since they're only rendered once when entering a route segment.
- Template files are nested and re-rendered, but they A) Don't fit semantically. B) Don't have access to route params. C) Are (just like layouts) executed/rendered from bottom to top.
- Middleware afaik is the default and (right now) recommended way for handling route guarding, but parsing route params manually is tedious and get's complicated really fast. Also, this just separates some routing functionality from the rest of the file-based routing.
- Guard functions (in the way implemented here) can be used to do multi-level guarding checks for you UI with no/minimal code redundancy but all the benefits from file based routing.
- TLDR: You could call this the return of the nested middleware, but in RSC rendering context and with the same API as a page or layout file.
- Update: @karlhorky  May also be a solution for many cases listed in https://github.com/vercel/next.js/discussions/46722 if people run the guards in the Node.js runtime (although guards don't support the normal middleware Request/Response API).

## How?

- Guard functions are the default export of `guard.js` files. Thus, they are loaded into the app rendering tree just like "page". "layout", "loading", etc. files.
- `createComponentTree` fetches the provided function from the guard.js file for every route (if existing) and appends it to an array of guard functions provided by the parent route segments.
- If a page should be rendered and any route guards apply, `createComponentTree` renders a newly introduced internal react component `RouteGuard` just before the page component is rendered.
- The `RouteGuard` component does nothing except executing all given guard functions (passed in as props) in a "parent->child" order.

## TBD
- Tests are still missing. An example is already given.
- Route guards currently don't get called at all when calling an invalid route. This _can_ be the target behavior, but I think I would prefer guards matching everything in a segment, even if the route is invalid (i.e. not logged-in users should just always be redirected away from /personal/...).
  - Tried to add RouteGuard to the NotFound comp tree in createComponentTree but that didn't work. Would have to take a look at that again.
- Don't know about runtime config yet. Since the file is getting parsed just like page files, having a runtime config in here should also work as expected?? Haven't tested that yet.
- Docs are there but short and the pretty images are missing.

### CODE REVIEW

**Comment by Vinlock** [comment]
Is this essentially like route middleware? 🤔

**Comment by glenntws** [comment]
You could call it that, although the guards don't have direct Request/Response capabilities as they run in RSC context.

It's more like a nested pre-"page render" logic that get's executed before the actual page render itself.

Example from my project:
- /app/[teamId]/guard.ts checks if the given teamId is valid for the user and otherwise redirects.
- /app/[teamId]/(only-if-configured)/guard.ts checks if configured:true on the selected team, otherwise redirects to the setup assistant.
  - I've now covered the whole guarding aspect of all team pages in the app instead of doing it repeatedly in page.tsx or parse urls manually in middleware.ts

**Comment by glenntws** [comment]
Any initial thoughts on this proposal from the maintainers? @leerob @ijjk

**Comment by karlhorky** [comment]
@glenntws another reason for this PR is that Middleware doesn't currently support Node.js APIs:

- https://github.com/vercel/next.js/discussions/46722

**Comment by tom-sherman** [comment]
Can you expand on the benefits this provides over calling `await myGuard()` at the top of the page component?

IMO the argument for wiring the props isn't strong enough because that can be done in userland in a more direct way with something like `export default guard(myGuard, Page);`

**Comment by glenntws** [comment]
> Can you expand on the benefits this provides over calling `await myGuard()` at the top of the page component?
> 
> IMO the argument for wiring the props isn't strong enough because that can be done in userland in a more direct way with something like `export default guard(myGuard, Page);`

This is exactly what I'm doing right now, and this massively increases the redundancy of the page guarding.

- If you have a guarded section with 3 levels of 2-3 child pages, you suddenly have the same guard call in every page component dozens of times across your whole app.
- Manual stacking: Creating a guard in a child page of an already guarded route means that you either have to call the Parent Guard manually or have to do the whole guarding again in the function.
- Easy to accidentally forget: You add a new page 2 months later. You forgot about the guard, so you now have one unguarded page along with 20 guarded ones. And since the page just works fine for you, you may even forget it, if people don't look closely in the review.

TLDR:
Stacking and integrating guard.ts functions manually in every page means:
- One additional import and line in every guard.ts file to include a parent guard (along with manually finding the closest guard)
- One additional import and line in every single page.tsx file in all child segments where at least one guard is active. And if you forget it, that page simply is unguarded.

**Comment by SukkaW** [comment]
It seems that the `guard.js` file convention proposed in this PR can only be used with RSC.

Ideally, the built-in guard from Next.js should function across Client-side Rendering, SSR, and RSC.

And here is what I am doing currently:

```
const GuardedComponent = () => {
  const userSession = useUser();
  // guard() will throw a `NoPermissionError` when the user doesn't have the permission
  guard(userSession, permissioinListFromSomewhere);
}

// We can implement the `HideWhenNoPermissonBoundary` component based on React Error Boundary
// When `HideWhenNoPermissonBoundary` receives `NoPermissionError`, it returns null
// When `HideWhenNoPermissonBoundary` receives the other errors, re-throw
<Suspense>
  <HideWhenNoPermissonBoundary>
    <GuardedComponent />
  </HideWhenNoPermissonBoundary>
</Suspense>

// We can also implement the `RedirectWhenNoPermissionBoundary` component, also based on React Error Boundary
// When `RedirectWhenNoPermissionBoundary` receives `NoPermissionError`, it calls `redirect()`
// When `RedirectWhenNoPermissionBoundary` receives the other errors, re-throw the error
<Suspense>
  <RedirectWhenNoPermissionBoundary to="/login">
    <GuardedComponent />
  </RedirectWhenNoPermissionBoundary>
</Suspense>
<Suspense>
  <RedirectWhenNoPermissionBoundary to="/403">
    <GuardedComponent />
  </RedirectWhenNoPermissionBoundary>
</Suspense>
```

React Error Boundary operates both on the server side and the client side. Moreover, we can position these boundaries wherever we prefer.

**Comment by glenntws** [comment]
> It seems that the `guard.js` file convention proposed in this PR can only be used with RSC + Middleware.
> 
> Ideally, the built-in guard from Next.js should function across Client-side Rendering, SSR, and RSC.
> 
> Also, such functionality should not be implemented at the route segment level. E.g. the page itself can be viewed by many roles, but some components (like a button or a tooltip) are only available for the admin role.
> 
> Here is what I am doing currently:
> 
> ```tsx
> const GuardedComponent = () => {
>   const userSession = useUser();
>   // guard() will throw a `NoPermissionError` when the user doesn't have the permission
>   guard(userSession, permissioinListFromSomewhere);
> }
> 
> // We can implement the `HideWhenNoPermissonBoundary` component based on React Error Boundary
> // When `HideWhenNoPermissonBoundary` receives `NoPermissionError`, it returns null
> // When `HideWhenNoPermissonBoundary` receives the other errors, re-throw
> <Suspense>
>   <HideWhenNoPermissonBoundary>
>     <GuardedComponent />
>   </HideWhenNoPermissonBoundary>
> </Suspense>
> 
> // We can also implement the `RedirectWhenNoPermissionBoundary` component, also based on React Error Boundary
> // When `RedirectWhenNoPermissionBoundary` receives `NoPermissionError`, it calls `redirect()`
> // When `RedirectWhenNoPermissionBoundary` receives the other errors, re-throw the error
> <Suspense>
>   <RedirectWhenNoPermissionBoundary to="/login">
>     <GuardedComponent />
>   </RedirectWhenNoPermissionBoundary>
> </Suspense>
> <Suspense>
>   <RedirectWhenNoPermissionBoundary to="/403">
>     <GuardedComponent />
>   </RedirectWhenNoPermissionBoundary>
> </Suspense>
> ```
> 
> And here is how `RedirectWhenNoPermissionBoundary` and `HideWhenNoPermissonBoundary` works:
> 
> React Error Boundary operates both on the server side and the client side, and we can put these boundaries wherever we prefer (not only the entire route segment but also per component).
> 
> During Server-Side Rendering (SSR), if an error is thrown by the `<GuardedComponent />`, React DOM Server will find the nearest `<Suspense />` boundary and emits its fallback in the HTML. During the hydration and client-side rendering, the error thrown by the `<GuardedComponent />` will be caught by the nearest Error Boundary where we can add redirection or hide by `return null` there.

This however again would not support automated stacking of guard's right?

**Comment by leerob** [comment]
Hey there! Thank you for this suggestion and for putting the work together to bring ideas.

[We're exploring a similar solution to the same problem](https://github.com/vercel/next.js/discussions/46722#discussioncomment-10262088) and will be hopefully soon publishing more information on where we're headed. For now, following along on that discussion would be the best place.

Thank you again 🙏
